sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -673,4 +786,81 @@ private void removeOne(String filename) throws IOException {
       }
     }
   }
+
+  /**
+   * Implementations create instances of {@link OutputWrapper} used by
+   * {@link DecoratedFileSink} and related classes to allow <em>decorating</em>, or otherwise
+   * transforming, the raw data that would normally be written directly to the {@link OutputStream}
+   * passed into {@link OutputWrapperFactory#create(OutputStream)}.
+   */
+  public interface OutputWrapperFactory extends Serializable {
+    /**
+     * @param channel the {@link WritableByteChannel} to wrap and allow for decorated, or otherwise
+     *        transformed, data to
+     * @return the {@link OutputWrapper} decorating, or otherwise transforming, the raw data
+     *         before writing it to {@code out}
+     * @throws IOException
+     */
+    public OutputWrapper create(WritableByteChannel channel) throws IOException;
+
+    /**
+     * @return the MIME type that should be used for the files that will hold the output data
+     * @see MimeTypes
+     * @see <a href=
+     *      'http://www.iana.org/assignments/media-types/media-types.xhtml'>http://www.iana.org/assignments/media-types/media-types.xhtml</a>
+     */
+    public String getMimeType();
+
+    /**
+     * @return an optional filename suffix, eg, ".gz" is returned by {@link CompressionType#GZIP}
+     */
+    public String getFilenameSuffix();
+  }
+
+  /**
+   * Subclasses of this should <em>decorate</em>, or otherwise transform, the raw data that would
+   * normally be written directly to the {@link WritableByteChannel} passed into
+   * {@link OutputWrapperFactory#create(WritableByteChannel)}. Implementations can
+   * accomplish this by creating a new {@link WritableByteChannel}, for instance, wrapped with an
+   * {@link OutputStream} providing the desired functionality, and passing it into
+   * {@link OutputWrapper#WriterOutputDecorator(WritableByteChannel)}.
+   * <p>
+   * The {@link #finish()} hook is provided for any logic needed to <em>finish</em> writing to the
+   * underlying stream without closing it.
+   *
+   * @see GZIPOutputStream#finish()
+   */
+  public static class OutputWrapper implements WritableByteChannel {
+    protected final WritableByteChannel channel;
+
+    public OutputWrapper(final WritableByteChannel channel) {
+      this.channel = channel;
+    }
+
+    /**
+     * Provided for any logic needed to <em>finish</em> writing to the {@link WritableByteChannel}
+     * without closing the underlying stream. This gets called after any <em>footer</em> data is
+     * written by the underlying {@link FileBasedWriter}.
+     * <p>
+     * Default implementation is a no-op.
+     *
+     * @throws IOException
+     */
+    public void finish() throws IOException {}
```

Comment:
when is this necessary? Seems like it would only ever be called right as you `close()`.